infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/DockerInternalRuntime.java

```diff
@@ -223,6 +267,79 @@ private void startAgents(String machineName, DockerMachine dockerMachine) throws
         }
     }
 
+    private void checkServersReadiness(String machineName, DockerMachine dockerMachine)
+            throws InternalInfrastructureException {
+        for (Map.Entry<String, ServerImpl> serverEntry : dockerMachine.getServers().entrySet()) {
+            String serverRef = serverEntry.getKey();
+            ServerImpl server = serverEntry.getValue();
+
+            LOG.info("Checking server {} of machine {}", serverRef, machineName);
+            checkServerReadiness(machineName, serverRef, server.getUrl());
+        }
+    }
+
+    private void checkServerReadiness(String machineName,
+                                      String serverRef,
+                                      String serverUrl)
+            throws InternalInfrastructureException {
+
+        if (!livenessChecksPaths.containsKey(serverRef)) {
+            return;
+        }
+        String livenessCheckPath = livenessChecksPaths.get(serverRef);
+        URL url;
+        try {
+            url = UriBuilder.fromUri(serverUrl)
+                            .replacePath(livenessCheckPath)
+                            .build()
+                            .toURL();
+        } catch (MalformedURLException e) {
+            // server URL is invalid, should not happen, skip server
```

Comment:
I'm not going to leave it here forever, but I want to make it more tolerant to mistakes while we are hard coding checks and haven't defined correct behavior in different situations.